tools: Add script to do split build in one command

seemethere opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141359

Usage:
```bash
python3 tools/packaging/split_wheel.py bdist_wheel
python3 tools/packaging/split_wheel.py install
python3 tools/packaging/split_wheel.py develop
```
Ideally this should make it easier to do the split build locally while
we're doing development.

Signed-off-by: Eli Uriegas <eliuriegas@meta.com>

seemethere committed:
tools: Add script to do split build in one command

Usage:

> python3 tools/build/wheel/split_wheel.py bdist_wheel
> python3 tools/build/wheel/split_wheel.py install
> python3 tools/build/wheel/split_wheel.py develop

Ideally this should make it easier to do the split build locally while
we're doing development.

Signed-off-by: Eli Uriegas <eliuriegas@meta.com>

[ghstack-poisoned]

seemethere committed:
Update on "tools: Add script to do split build in one command"


Usage:
```bash
python3 tools/build/wheel/split_wheel.py bdist_wheel
python3 tools/build/wheel/split_wheel.py install
python3 tools/build/wheel/split_wheel.py develop
```
Ideally this should make it easier to do the split build locally while
we're doing development.

Signed-off-by: Eli Uriegas <eliuriegasmeta.com>

[ghstack-poisoned]

seemethere commented:
@pytorchbot label "topic: not user facing"

kit1980 commented on tools/build/wheel/split_wheel.py:
```diff
@@ -0,0 +1,128 @@
+        extra_env={"BUILD_LIBTORCH_WHL": "1", "BUILD_PYTHON_WHEEL": "0"},
+    )
+    # Equivalent to running the following:
+    #     > BUILD_LIBTORCH_WHL=0 BUILD_PYTHON_ONLY=1 python setup.py bdist_wheel --cmake
```
`BUILD_PYTHON_ONLY` or `BUILD_PYTHON_WHEEL`?

seemethere replied:
You're right this is a typo on my end, I expected them to be consistent but they were not

kit1980 commented on tools/build/wheel/split_wheel.py:
```diff
@@ -0,0 +1,128 @@
+        str: The path to the Git repository root.
+    """
+    try:
+        # Run the 'git rev-parse --show-toplevel' command
```
Do you need this comment? It just repeats what's in the code and will need to be maintained to be in sync.

seemethere replied:
Removed!

seemethere committed:
Update on "tools: Add script to do split build in one command"


Usage:
```bash
python3 tools/build/wheel/split_wheel.py bdist_wheel
python3 tools/build/wheel/split_wheel.py install
python3 tools/build/wheel/split_wheel.py develop
```
Ideally this should make it easier to do the split build locally while
we're doing development.

Signed-off-by: Eli Uriegas <eliuriegasmeta.com>

[ghstack-poisoned]

malfet commented on tools/build/wheel/split_wheel.py:
```diff
@@ -0,0 +1,128 @@
+# TODO: Add a function to do submodule setup
+
+
+def get_git_root() -> Path:
```
You need project root, rather than a git root (as this command should ideally work if someone downloads our release tarballs), I.e. it should be computed as `PATH(__file__).parent.parent.parent.parent`

malfet commented on tools/build/wheel/split_wheel.py:
```diff
@@ -0,0 +1,128 @@
+
+
+def get_setup_py() -> Path:
+    return get_git_root() / Path("setup.py")
```
Nit
```suggestion
    return get_git_root() / "setup.py"
```

malfet commented on tools/build/wheel/split_wheel.py:
```diff
@@ -0,0 +1,128 @@
+        extra_env = {}
+    cmd = [sys.executable, str(get_setup_py()), *cmd_args]
+    logger.debug("+ %s", " ".join(cmd))
+    subprocess.run(
+        cmd,
+        # Give the parent environment to the subprocess
+        env={**os.environ, **extra_env},
+        check=True,
+    )
```
```suggestion
    subprocess.check_call(
        cmd,
        # Give the parent environment to the subprocess
        env={**os.environ, **extra_env},
    )
```

seemethere replied:
`subprocess.run(check=true)` is actually the supported version of this now.

From: https://github.com/python/cpython/blob/75ef92da291247fd9d21ec26bd22bf8055b5f330/Lib/subprocess.py#L16-L40

malfet commented on tools/build/wheel/split_wheel.py:
```diff
@@ -0,0 +1,128 @@
+def bdist_wheel():
+    # Equivalent to running the following:
+    #     > BUILD_LIBTORCH_WHL=1 BUILD_PYTHON_ONLY=0 python setup.py bdist_wheel
+    logger.info("Building libtorch wheel")
+    setup_py(
+        ["bdist_wheel"],
+        extra_env={"BUILD_LIBTORCH_WHL": "1", "BUILD_PYTHON_ONLY": "0"},
+    )
+    # Equivalent to running the following:
+    #     > BUILD_LIBTORCH_WHL=0 BUILD_PYTHON_ONLY=1 python setup.py bdist_wheel --cmake
+    logger.info("Building torch wheel")
+    setup_py(
+        ["bdist_wheel", "--cmake"],
+        extra_env={"BUILD_LIBTORCH_WHL": "0", "BUILD_PYTHON_ONLY": "1"},
+    )
```
Avoid code repetition (and it feels like you want to refactor it in a helper function, don't you?)
```suggestion
    for build_python in ["0", "1"]:
        logger.info(f'Building {"libtorch" if build_python == "0" else "torch wheel"}')
      setup_py(
          ["bdist_wheel"],
          extra_env={"BUILD_LIBTORCH_WHL": "1", "BUILD_PYTHON_ONLY": build_python},
      )
```

seemethere replied:
Should mostly be DRY now!

malfet reviewed (commented):
Looks like `bdist_wheel()`, `install` and `develop` are almost exact copies of each other and would benefit from helper `run_split_build_cmd` function

seemethere committed:
Update on "tools: Add script to do split build in one command"


Usage:
```bash
python3 tools/build/wheel/split_wheel.py bdist_wheel
python3 tools/build/wheel/split_wheel.py install
python3 tools/build/wheel/split_wheel.py develop
```
Ideally this should make it easier to do the split build locally while
we're doing development.

Signed-off-by: Eli Uriegas <eliuriegasmeta.com>

[ghstack-poisoned]

clee2000 commented on tools/build/wheel/split_wheel.py:
```diff
@@ -0,0 +1,103 @@
+            "Requirements not installed, run the following command to install:"
+        )
+        logger.error(
+            "    > %s install -r %s/requirements.txt", sys.executable, ROOT_PATH
```
If I'm understanding correctly, this is going to print `python install -r <root path>/requirements.txt`, should it be pip instead?

seemethere replied:
Good catch! Changing to `python -m pip`!

seemethere replied:
Yup! Should be improved in the latest iteration to `python -m pip`

clee2000 commented on tools/build/wheel/split_wheel.py:
```diff
@@ -0,0 +1,103 @@
+    logger.info("Running %s for libtorch wheel", cmd)
+    setup_py(
+        [cmd],
+        extra_env={"BUILD_LIBTORCH_WHL": "1", "BUILD_PYTHON_ONLY": "0"},
```
Does it need to inherit any other environment variables?  Like the bash script and build ymls set some env vars, are any of those needed?

seemethere replied:
Those should be passed above in the setup_py command through `**os.environ`

seemethere committed:
Update on "tools: Add script to do split build in one command"


Usage:
```bash
python3 tools/build/wheel/split_wheel.py bdist_wheel
python3 tools/build/wheel/split_wheel.py install
python3 tools/build/wheel/split_wheel.py develop
```
Ideally this should make it easier to do the split build locally while
we're doing development.

Signed-off-by: Eli Uriegas <eliuriegasmeta.com>

[ghstack-poisoned]

seemethere committed:
Update on "tools: Add script to do split build in one command"


Usage:
```bash
python3 tools/build/wheel/split_wheel.py bdist_wheel
python3 tools/build/wheel/split_wheel.py install
python3 tools/build/wheel/split_wheel.py develop
```
Ideally this should make it easier to do the split build locally while
we're doing development.

Signed-off-by: Eli Uriegas <eliuriegasmeta.com>

[ghstack-poisoned]

seemethere commented:
@pytorchbot merge
